src/inventory/common.py

```diff
@@ -35,4 +55,119 @@ def slot_has_item(slot_img: np.ndarray) -> bool:
+        green_mask, _ = color_filter(img_post, Config().colors["green_slot"])
+
+        blue_red_mask = np.bitwise_or(blue_mask, red_mask)
+        final = np.bitwise_and.reduce([blue_red_mask, green_mask, diff_thresh])
+        _, roi = trim_black(final)
+        return roi
+    except BaseException as err:
+        Logger.error(f"_calc_item_roi: Unexpected {err=}, {type(err)=}")
+        return None
+
+def tome_state(img: np.ndarray = None, tome_type: str = "tp", roi: list = None):
+    img = img if img is not None else grab()
+    if (tome_found := TemplateFinder().search([f"{tome_type.upper()}_TOME", f"{tome_type.upper()}_TOME_RED"], img, roi = roi, threshold = 0.8, best_match = True, normalize_monitor = True)).valid:
+        if tome_found.name == f"{tome_type.upper()}_TOME":
+            state = "ok"
+        else:
+            state = "empty"
+        position = tome_found.center
+    else:
+        state = position = None
+    return state, position
+
+def id_item_with_tome(item_location: list, id_tome_location: list):
+    mouse.move(id_tome_location[0], id_tome_location[1], randomize=4, delay_factor=[0.4, 0.8])
+    wait(0.2, 0.4)
+    mouse.click(button="right")
+    wait(0.1)
+    mouse.move(item_location[0], item_location[1], randomize=4, delay_factor=[0.4, 0.8])
+    wait(0.1)
+    mouse.click(button="left")
+    consumables.increment_need("id", 1)
+    wait(0.2, 0.4)
+
+def transfer_items(items: list, action: str = "drop") -> list:
+    #requires open inventory / stash / vendor
+    img = grab()
+    filtered = []
+    left_panel_open = detect_screen_object(ScreenObjects.LeftPanel, img).valid
+    if action == "drop":
+        filtered = [ item for item in items if item.keep == False and item.sell == False ]
+    elif action == "sell":
+        filtered = [ item for item in items if item.keep == False and item.sell == True ]
+        if not left_panel_open:
+            Logger.error(f"transfer_items: Can't perform, vendor is not open")
+    elif action == "stash":
+        if detect_screen_object(ScreenObjects.GoldBtnStash, img).valid:
+            filtered = [ item for item in items if item.keep == True ]
+        else:
+            Logger.error(f"transfer_items: Can't perform, stash is not open")
+    else:
+        Logger.error(f"transfer_items: incorrect action param={action}")
+    if filtered:
+        # if dropping, control+click to drop unless left panel is open, then drag to middle
+        # if stashing, control+click to stash
+        # if selling, control+click to sell
+        # TODO: if purchasing, right-click to buy
+        # TODO: if purchasing stack, shift+right-click to buy
+        if (action == "drop" and not left_panel_open) or action in ["sell", "stash"]:
+            keyboard.send('ctrl', do_release=False)
+            wait(0.2, 0.4)
+        for item in filtered:
+            attempts = 0
+            while True:
+                mouse.move(*item.pos, randomize=4, delay_factor=[0.2, 0.4])
```

Comment:
Can extract a function or two here

Reply:
Thanks for looking things over! Really appreciate it.

I know that I *could* extract functions here, but I generally prefer to not extract things into separate functions if I'm sufficiently distant from the base scripts (game_controller, main, bot, etc.) and if I don't plan to use those functions anywhere else now or in the future, mostly to improve readability. I don't think the contents of this function in particular will need to be used elsewhere, so I'll probably leave them here.

Reply:
It's more about supporting this by another person some time later. When I see a good bunch of mouse & keyboard interactions in the same function I'm completely clueless what's going on and it's hard to reason what's going on without seeing it running. And it's located in the loop as well. A functions extracted in a form of "_hover_over_XX", "_drag_xxx_to_yyy", "_select_zzz" would boost readability of this code significantly without really any drawbacks